staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -606,26 +618,36 @@ func getNewItemFunc(listObj runtime.Object, v reflect.Value) func() runtime.Obje
 	}
 }
 
-func (s *store) Count(ctx context.Context, key string) (int64, error) {
-	preparedKey, err := s.prepareKey(key)
+func (s *store) Stats(ctx context.Context) (stats storage.Stats, err error) {
+	if s.stats != nil {
+		return s.stats.Stats(ctx)
+	}
+	startTime := time.Now()
+	count, err := s.client.Kubernetes.Count(ctx, s.pathPrefix, kubernetes.CountOptions{})
+	metrics.RecordEtcdRequest("listWithCount", s.groupResource, err, startTime)
 	if err != nil {
-		return 0, err
+		return storage.Stats{}, err
 	}
+	return storage.Stats{
+		ObjectCount: count,
+	}, nil
+}
 
-	// We need to make sure the key ended with "/" so that we only get children "directories".
-	// e.g. if we have key "/a", "/a/b", "/ab", getting keys with prefix "/a" will return all three,
-	// while with prefix "/a/" will return only "/a/b" which is the correct answer.
-	if !strings.HasSuffix(preparedKey, "/") {
-		preparedKey += "/"
+func (s *store) SetKeysFunc(keys storage.KeysFunc) {
+	if s.stats != nil {
+		s.stats.SetKeysFunc(keys)
 	}
+}
 
+func (s *store) getKeys(ctx context.Context) ([]string, error) {
 	startTime := time.Now()
-	count, err := s.client.Kubernetes.Count(ctx, preparedKey, kubernetes.CountOptions{})
-	metrics.RecordEtcdRequest("listWithCount", s.groupResource, err, startTime)
-	if err != nil {
-		return 0, err
+	resp, err := s.client.KV.Get(ctx, s.pathPrefix, clientv3.WithPrefix(), clientv3.WithKeysOnly())
```

Comment:
@serathius @BenTheElder this appears to be a rework of some things that were changed in https://github.com/kubernetes/kubernetes/pull/125258

On the kine side, we had not previously implemented the `KeysOnly` option, as the apiserver was using `CountOnly` instead.

I fail to see how listing keys, and then counting them, is better than just asking etcd for a count of keys, is an improvement. Is this something that could be addressed in the future?

Reply:
Ah, I see that it's not JUST using it for counts, the keys are also used to prune the cache? That makes things more complicated.